packages/midi/src/index.ts

```diff
@@ -0,0 +1,170 @@
+import { Signal } from './Signal.ts'
+
+/**
+ * Common MIDI message types as human readable names.
+ */
+export enum MidiMessageType {
+  Clock = 'Clock',
+  Start = 'Start',
+  Continue = 'Continue',
+  Stop = 'Stop',
+  NoteOff = 'Note Off',
+  NoteOn = 'Note On',
+  PolyphonicKeyPressure = 'Polyphonic Key Pressure (Aftertouch)',
+  ControlChange = 'Control Change',
+  ProgramChange = 'Program Change',
+  ChannelPressure = 'Channel Pressure (Aftertouch)',
+  PitchBendChange = 'Pitch Bend Change',
+  Unknown = 'Unknown',
+}
+
+/**
+ * A simple type that provides both the device and the message of a MIDI event.
+ */
+export type MIDIEvent = {
+  device: MIDIInput
+  message: MIDIMessageEvent
+}
+
+/**
+ * A class that handles MIDI input devices and messages.
+ */
+export class Midi {
+  /**
+   * The list of MIDI input devices connected to the system.
+   */
+  public devices: MIDIInput[] = []
+
+  /**
+   * The MIDI access object that allows for interaction with MIDI devices.
+   */
+  private midiAccess: MIDIAccess | null = null
+
+  /**
+   * A map of MIDI input devices to their event listeners.
+   */
+  private eventListeners: Map<MIDIInput, (event: MIDIMessageEvent) => void> = new Map()
+
+  /**
+   * A callback that is called when the list of MIDI devices changes.
+   */
+  public onDeviceChange: Signal<void> = new Signal<void>()
+
+  /**
+   * A callback that is called when a MIDI message is received.
+   */
+  public onMidiMessage: Signal<MIDIEvent> = new Signal<MIDIEvent>()
+
+  constructor() {
+    this.findMidiDevices()
+  }
+
+  /**
+   * Clears all MIDI event listeners from the devices.
+   */
+  private clearMidiEventListeners = (): void => {
+    this.devices.forEach((device) => {
+      const listener = this.eventListeners.get(device)
+      if (listener) {
+        device.removeEventListener('midimessage', listener)
+        this.eventListeners.delete(device)
+      }
+    })
+  }
```

Comment:
We probably want this method to be public. Or wrapped in some `destroy` sort of method. e.g. needed in React when a component unmounts.

Reply:
I have used a stripped down version of your Midi class in Clock, but `clearMidiEventListeners` is now async and has some extra promise logic. The reason behind this, is that if someone created the class and then tried to clear the listeners, it doesn't work because the listeners are not created yet (as the device finding is async). This might seem like a weird edge case but actually in React when you're in dev mode, `useEffect` is called twice in succession as a way to force you to be clean with your code.

Here's how I solved it:
https://github.com/nudibranchrecords/hedron/pull/492/files#diff-b6e106463b54d5e5622bd46ba9118fd94fa768262c379eebb4df919003409fea